extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,309 @@
+  }
+
+  /**
+   * Generic rejection handler for amp-user-notification.
+   * @private
+   */
+  onRejection_(msg) {
+    console/*OK*/.error('Rejection ocurred on `amp-user-notification` ' +
+        `with html element id "${this.elementId_}". message: ${msg}`);
+  }
+
+  /**
+   * Get async cid service.
+   * @private
+   */
+  getAsyncCid_() {
+    return cidFor(this.win_).then((cid) => {
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   * @export
+   */
+  show() {
+    this.element.classList.add('amp-active');
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   * @export
+   */
+  dismiss() {
+    this.vsync_.mutate(() => {
+      this.element.classList.remove('amp-active');
+      this.element.classList.add('amp-hidden');
+      this.dialogResolve_();
+    });
+  }
+}
+
+
+/**
+ * UserNotification service that is a per window singleton on installation.
+ */
+class UserNotification {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<UserNotificationDefer>} */
+    this.deferRegistry_ = Object.create(null);
+
+    this.glasspane = this.win_.document.createElement('div');
+    this.glasspane.classList.add('-amp-glasspane');
+    this.win_.document.body.appendChild(this.glasspane);
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @return {!Promise|null}
+   */
+  get(id) {
+    let defer = this.getElementDeferById_(id);
+    if (defer) {
+      return defer.promise;
+    }
+    return null;
+  }
+
+  /**
+   * Checks the internal cache if an amp-user-notification component
+   * has been registered with the element id given.
+   * @param {string|undefined} id
+   * @return boolean
+   */
+  has(id) {
+    if (!id) {
+      return false;
+    }
+    return !!this.getElementDeferById_(id);
+  }
+
+  /**
+   * Register an instance of `amp-user-notification`.
+   * @param {string} id html element id of the amp-user-notification component.
+   * @param {!Promise} consentPromise promise to be resolved when a human
+   *     accepts the terms.
+   */
+  registerUserNotification(id, consentPromise) {
+    let deferred = this.getElementDeferById_();
+    let resolve_;
+
+    assert(this.getElementDeferById_(id) == undefined, 'Duplicate ' +
+        'amp-use-notification id found.');
+
+    const promise = new Promise((resolve, reject) => {
+      resolve_ = resolve;
+    });
+
+    deferred = this.deferRegistry_[id] = {
+      promise: promise,
+      resolve: resolve_
+    }
+    deferred.resolve(consentPromise);
+  }
+
+  /**
+   * Retrieve UserNotificationDefer object.
+   * @param {string} id
+   * @return {UserNotificationDefer|undefined}
```

Comment:
since `UserNotificationDefer` is a typedef to a record type, they are inherently non nullable. (correct me if im wrong)